### PR TITLE
[RFC]topology: hda: add a volume component to convert 24bit format

### PR DESCRIPTION
**Pull request by RanderWang**
Fix the bug: sof audio pipeline with the format of 24bit doesn't
work on HDA platform. It is caused by sample data layout. SOF supports
S24_LE, and each sample format is like pattern of 0ddd, d stands for
data. Samples are padded with 0‟s at the MSB.

Link dma copys samples directly to hda bus, but hda bus requires samples
layout like pattern of ddd0. Samples are padded with 0‟s at the LSB to
left justify the sample within the container.

Suggesting by Liam and Seppo, using volume to convert S24_LE to S32_LE
in topology file.

This fixes https://github.com/thesofproject/sof/issues/700
Signed-off-by: Rander Wang <rander.wang@linux.intel.com>

### CODE REVIEW

**Comment by singalsu** [comment]
Hi, I wonder why the volume component is instantiated by DAI for the HDA type and not by the playback and capture pipelines? It makes the HDA DAI different from others and creates  risk for unneccessary volume components duplication. Did I interpret the patch m4 code correctly?

**Comment by RanderWang** [comment]
> Hi, I wonder why the volume component is instantiated by DAI for the HDA type and not by the playback and capture pipelines? It makes the HDA DAI different from others and creates risk for unnecessary volume components duplication. Did I interpret the patch m4 code correctly?

At here, I don't know this playback or capture pipeline is for HDA or not, so I can't add a volume here.
Here I add a implicit volume here to let user doesn't take care of HW detail. And if we let users know this limitation and do it by themselves, they could use volume optimally

**Comment by RanderWang** [comment]
> This means that all HDA pipes must have a volume ? Why not just use the sof volume pipeline on the HDA pipes that need them.

No, only the hda pipes with 24bit dai.  And if there is a 24bit dai, the hda pipe needs a sof volume. I add no redundant volume except mentioned by Seppo

**Comment by ranj063** [comment]
> > Hi, I wonder why the volume component is instantiated by DAI for the HDA type and not by the playback and capture pipelines? It makes the HDA DAI different from others and creates risk for unnecessary volume components duplication. Did I interpret the patch m4 code correctly?
> 
> At here, I don't know this playback or capture pipeline is for HDA or not, so I can't add a volume here.
> Here I add a implicit volume here to let user doesn't take care of HW detail. And if we let users know this limitation and do it by themselves, they could use volume optimally

@RanderWang this change is too difficult to follow. 
I think the simplest change is as Liam suggested. When you create the top-level m4 file, you know if it is for a HDA pipeline, so you can just use sof-volume-playback/capture for those.

File: tools/topology/m4/dai.m4 [hunk 1]
```diff
@@ -44,7 +44,8 @@ define(`W_DAI_OUT',
 `SectionVendorTuples."'N_DAI_OUT($2)`_tuples_comp_str" {'
 `	tokens "sof_comp_tokens"'
 `	tuples."string" {'
-`		SOF_TKN_COMP_FORMAT'	STR($4)
+`		SOF_TKN_COMP_FORMAT'    `ifelse($1, "HDA",
+		`ifelse($4, `s24le', `"s32le"', STR($4))', `STR($4)')'
```
Comment (ranj063) [inline tools/topology/m4/dai.m4]: @RanderWang why is this needed? Why are you overriding the s24le set in the top-level m4 to s32le?

Comment (ranj063) [reply tools/topology/m4/dai.m4]: Doesnt this change make the other changes redundant?

**Comment by ranj063** [comment]
> @ranj063 please have a look at #700. it is a limitation of hda dma, and it should be transparent to user

@RanderWang if you want to make it transparent, you need to have the change in the top-level m4. Besides, user is not the end user here right? It is someone like you or me who is quite aware of the limitation. I feel like this is a roundabout way of addressing the problem. The simpler solution is to add a comment before the PIPELINE_PCM_ADD for HDA pipelines about why they need to use the volume pipelines.

File: tools/topology/m4/dai.m4 (hunk 1 above)
Comment (RanderWang) [reply tools/topology/m4/dai.m4]: volume component in FW would check source and sink format to select conversion function, so
I need to change dai format. 

Yes, you are right if there is a volume at the end of pipeline. I am try to figure out how to know whether there is a volume at the end of pipeline and if I know it, I don't need to add volume again

**Comment by RanderWang** [comment]
> > @ranj063 please have a look at #700. it is a limitation of hda dma, and it should be transparent to user
> 
> @RanderWang if you want to make it transparent, you need to have the change in the top-level m4. Besides, user is not the end user here right? It is someone like you or me who is quite aware of the limitation. I feel like this is a roundabout way of addressing the problem. The simpler solution is to add a comment before the PIPELINE_PCM_ADD for HDA pipelines about why they need to use the volume pipelines.

Yes, your idea is the simplest one, and I would get all feedback from others.
Here, users maybe other persons in community, so they may be confused about this.

**Comment by ranj063** [comment]
> > > @ranj063 please have a look at #700. it is a limitation of hda dma, and it should be transparent to user
> > 
> > 
> > @RanderWang if you want to make it transparent, you need to have the change in the top-level m4. Besides, user is not the end user here right? It is someone like you or me who is quite aware of the limitation. I feel like this is a roundabout way of addressing the problem. The simpler solution is to add a comment before the PIPELINE_PCM_ADD for HDA pipelines about why they need to use the volume pipelines.
> 
> Yes, your idea is the simplest one, and I would get all feedback from others.
> Here, users maybe other persons in community, so they may be confused about this.

@RanderWang how baout adding a debug check in debug.m4 for HDA DAI's which have s24le pipeline format?

**Comment by RanderWang** [comment]
> > 
> > Yes, your idea is the simplest one, and I would get all feedback from others.
> > Here, users maybe other persons in community, so they may be confused about this.
> 
> @RanderWang how about adding a debug check in debug.m4 for HDA DAI's which have s24le pipeline format?

so if it is s24le, then do what ?

**Comment by RanderWang** [comment]
update my PR. Now check whether there is a volume at the end of pipeline. Don't add volume if there is one.

**Comment by RanderWang** [comment]
Now if it is HDA dai && format == S24le && no volume at the end of pipeline, a volume would be added.

**Comment by lgirdwood** [comment]
@RanderWang I still see this as the wrong cource as it add complexity where it's not needed. Can you not define pipelines for each required machine as
```
PIPELINE_PCM_ADD(sof/pipe-volume-playback.m4,
```

**Comment by RanderWang** [comment]
> @RanderWang I still see this as the wrong course as it adds complexity where it's not needed. Can you define pipelines for each required machine like
> 
> ```
> PIPELINE_PCM_ADD(sof/pipe-volume-playback.m4,
> ```

Done, thanks!

**Comment by lgirdwood** [comment]
Sorry @RanderWang I mean this
```diff
diff --git a/tools/topology/sof-whl-hda.m4 b/tools/topology/sof-whl-hda.m4
index ff6aa5f2..a4fa719a 100644
--- a/tools/topology/sof-whl-hda.m4
+++ b/tools/topology/sof-whl-hda.m4
@@ -28,7 +28,7 @@ include(`platform/intel/bxt.m4')
 
 # Low Latency playback pipeline 1 on PCM 0 using max 2 channels of s24le.
 # Schedule 48 frames per 1000us deadline on core 0 with priority 0
-PIPELINE_PCM_ADD(sof/pipe-passthrough-playback.m4,
+PIPELINE_PCM_ADD(sof/pipe-volume-playback.m4,
        1, 0, 2, s24le,
        48, 1000, 0, 0)
 
```
This we we include a volume transformation for each pipeline and it does not impact other HDA users who may not need or want a volume.

File: tools/topology/m4/dai.m4 (hunk 1 above)
Comment (lgirdwood) [reply tools/topology/m4/dai.m4]: @RanderWang I still see Intel specific logic in generic code. Why don't you take each pipeline in machine.m4 and use a volume pipeline(instead of passthrough) for every pipeline that has this requirement.  Please see the diff I commented yesterday.

**Comment by RanderWang** [comment]
@lgirdwood    sof-hda-generic.m4 is used by intel platform.  sof-whl-hda.m4 is not used now.
Do you mean: use sof-hda-whl.m4 for whl, sof-hda-icl.m4 for ICL,  sof-hda-apl.m4 for APL?

Thanks!

**Comment by lgirdwood** [comment]
@RanderWang I mean that sof-hda-generic.m4 will be for *all* HDA capable platforms and not just Intel platforms so must be generic (you cant make any Intel only changes to this file). You can however make Intel specific changes to intel files.  So dai.m4 is generic and should not contain Intel specific logic.

**Comment by RanderWang** [comment]
> @RanderWang I mean that sof-hda-generic.m4 will be for _all_ HDA capable platforms and not just Intel platforms so must be generic (you cant make any Intel only changes to this file). You can however make Intel specific changes to intel files. So dai.m4 is generic and should not contain Intel specific logic.
> 
> If you renamed sof-hda-generic.m4 to sof-hda-intel-common.m4 then there is no problem making these changes.

Yes, one problem is in our kernel code, the topology for intel is named sof-hda-generic.tplg. This is not
a complex issue. If we used sof-hda-intel-common.m4, we need to build the file to sof-hda-generic.tplg

**Comment by RanderWang** [comment]
Update PR, thanks!

**Comment by lgirdwood** [comment]
@RanderWang pls check CI

**Comment by RanderWang** [comment]
@lgirdwood   CI is successful.
